tunnel-server/src/ssh-server.ts

```diff
@@ -4,196 +4,265 @@ import net from 'net'
+              reject?.()
             }
           )
-        })
 
-        const socketPath = path.join(socketDir, `s_${clientId}_${randomBytes(16).toString('hex')}`)
+        })
+        .on('error', err => { preevySshClient.emit('error', err) })
+        .on('session', accept => {
+          log.debug('session')
+          const session = accept()
 
-        const closeSocketServer = () => socketServer.close()
+          session.on('exec', async (accept, reject, info) => {
+            log.debug('exec %j', info)
+            if (info.command !== 'hello') {
+              log.error('invalid exec command %j', info.command)
+              reject()
+              return
+            }
 
-        socketServer
-          .listen(socketPath, () => {
-            log.debug('streamlocal-forward@openssh.com: calling accept: %j', accept)
-            accept?.()
-            tunnels.add(requestedSocketPath)
-            onPipeCreated?.({clientId, remotePath: requestedSocketPath, localSocketPath: socketPath, publicKey, access: params.access})
-          })
-          .on('error', (err: unknown) => {
-            log.error('socketServer error: %j', err)
-            socketServer.close()
+            const channel = accept()
+            preevySshClient.emit('hello', {
+              stdout: channel.stdout,
+              exit: (status: number) => {
+                channel.stdout.exit(status)
+                if (tunnels.size === 0) {
+                  channel.close()
+                }
+              },
+            })
+            // channel.stdout.on('close', () => {
+            //   log.debug('hello close, tunnels: %d', tunnels.size)
+            //   if (tunnels.size === 0) {
+            //     channel.close()
+            //   }
+            // })
           })
-          .on('close', () => {
-            log.debug('socketServer close: %j', socketPath)
-            tunnels.delete(requestedSocketPath)
-            onPipeDestroyed?.({clientId, remotePath: requestedSocketPath, localSocketPath: socketPath})
-            client.removeListener('close', closeSocketServer)
-          })
-
-        client.once('close', closeSocketServer)
-      })
-      .on('session', (accept) => {
-        log.debug('session')
-        const session = accept()
-
-        session.on('exec', (accept, reject, info) => {
-          log.debug('exec %j', info)
-          if (info.command !== 'hello') {
-            log.error('invalid exec command %j', info.command)
-            reject()
-            return
-          }
-          const channel = accept()
-          log.info({clientId, tunnels: [...tunnels]}, 'client connected')
-          channel.stdout.write(onHello(clientId, [...tunnels]))
-          channel.stdout.exit(0)
-          if (tunnels.size === 0) {
-            channel.close()
-          }
         })
-      })
-  }
-)
+    }
+  )
+
+  return Object.assign(serverEmitter, {
```

Comment:
nit - might be clearer to return the server and assign the "on:" server Emitter.on
Since the return value is basically SSH server with an event handler.